server/src/main/java/com/objectcomputing/checkins/services/kudos/KudosServicesImpl.java

```diff
@@ -0,0 +1,289 @@
+                throw e;
+            }
+        });
+
+        return savedKudos;
+    }
+
+    @Override
+    public Kudos approve(Kudos kudos) {
+
+        if (!currentUserServices.isAdmin()) {
+            throw new PermissionException("You are not authorized to perform this operation");
+        }
+
+        Kudos existingKudos = kudosRepository.findById(kudos.getId()).orElseThrow(() ->
+                new BadArgException("Kudos with id %s does not exist"));
+
+        if (existingKudos.getDateApproved() != null) {
+            throw new BadArgException("Kudos with id %s has already been approved");
+        }
+
+        existingKudos.setDateApproved(LocalDate.now());
+
+        return kudosRepository.update(existingKudos);
+    }
+
+    @Override
+    public KudosResponseDTO getById(UUID id) {
+
+        Kudos kudos = kudosRepository.findById(id).orElseThrow(() ->
+                new NotFoundException("Kudos with id %s does not exist"));
+
+        UUID currentUserId = currentUserServices.getCurrentUser().getId();
+        boolean isSender = currentUserId.equals(kudos.getSenderId());
+
+        if (kudos.getDateApproved() == null) {
+            // If not yet approved, only admins and the sender can access the kudos
+            if (!currentUserServices.isAdmin() && !isSender) {
+                throw new PermissionException("You are not authorized to perform this operation");
+            }
+        } else {
+            // If approved, admins, the sender, and the recipient can access the kudos
+
+            List<KudosRecipient> recipients = kudosRecipientRepository.findByKudosId(id);
+            boolean isRecipient = recipients
+                    .stream()
+                    .anyMatch(recipient -> recipient.getMemberId().equals(currentUserId));
+
+            if (!currentUserServices.isAdmin() && !isSender && !isRecipient) {
+                throw new PermissionException("You are not authorized to perform this operation");
+            }
+        }
+
+        return constructKudosResponseDTO(kudos);
+
+    }
+
+    @Override
+    public Optional<Kudos> findById(UUID id) {
+        return kudosRepository.findById(id);
+    }
+
+    @Override
+    public boolean delete(UUID id) {
+        if (!currentUserServices.isAdmin()) {
+            throw new PermissionException("You are not authorized to do this operation");
+        }
+
+        Kudos kudos = kudosRepository.findById(id).orElseThrow(() ->
+                new NotFoundException("Kudos with id %s does not exist"));
+
+        // Delete all KudosRecipients associated with this kudos
+        List<KudosRecipient> recipients = kudosRecipientServices.getAllByKudosId(kudos.getId());
+        kudosRecipientRepository.deleteAll(recipients);
+
+        kudosRepository.deleteById(id);
+        return true;
+    }
+
+    @Override
+    public List<KudosResponseDTO> findByValues(@Nullable UUID recipientId, @Nullable UUID senderId, @Nullable Boolean isPending, @Nullable Boolean Public) {
```

Comment:
`Public` is a bad name for an argument, and it isn't used here in this method 🤔 

Is there a requirement to be able to search on both recipient AND sender? Or pending AND sender? Here we search for pending if isPending is set, and we will then ignore any senderId that is also passed?